### DISCUSSION TITLE
How can i send/receive/modify packet?

### DISCUSSION THREAD

**Discussion by giqnt** [Minestom/Minestom#1195] [Help] [answered]
How can i send/receive/modify packet?

**Comment by KrystilizeNevaDies** [comment]
Sending:
```java
var packet = new XXXPacket(...);
player.sendPacket(packet);
```

Receiving:
PlayerPacketEvent

Modifying:
Wdym by modifying?

**Reply by giqnt** [reply]
by modifying, i mean modifying incoming and outgoing packet.

**Reply by KrystilizeNevaDies** [reply]
We don't support modifying outgoing packets as that limits optimisations.

As for modifying incoming packets, you can use the PlayerPacketEvent for that.

**Reply by TheMode** [reply]
There are PRs adding support for it like https://github.com/Minestom/Minestom/pull/1141
But this is not something we are yet willing to add, at least not in this way

**Comment by VegetalDev** [comment, marked as answer]
You can send packets either directly to a player, a Viewable implementation using `Viewable#sendPacketToViewers`, or a PacketGroupingAudience with `PacketGroupingAudience#sendGroupedPacket`. PacketGroupingAudience and Viewable work similarly in that they both allow you to send packets to a group of players, which is often useful. 

For reading incoming packets, you should use PlayerPacketEvent, like this for example: 

```java
MinecraftServer.getGlobalEventHandler().addListener(PlayerPacketEvent.class, packetEvent -> {
    //do stuff
    //cancel packets using packetEvent.setCancelled(true);
});
```

However, as was already mentioned, you can't really modify outgoing (client-bound) packets, though you have full control of inbound (server-bound) packets.
